packages/common/src/TenderlySimulation.ts

```diff
@@ -0,0 +1,216 @@
+};
+
+const createRequestUrl = (tenderlyEnv: TenderlyEnvironment, fork?: ForkParams): string => {
+  const baseUrl = `https://api.tenderly.co/api/v1/account/${tenderlyEnv.user}/project/${tenderlyEnv.project}/`;
+  return fork === undefined ? baseUrl + "simulate" : baseUrl + "fork/" + fork.id + "/simulate";
+};
+
+const createRequestBody = (simulationParams: TenderlySimulationParams): TenderlyRequestBody => {
+  const body: TenderlyRequestBody = {
+    save: true,
+    save_if_fails: true,
+    simulation_type: "full",
+    network_id: simulationParams.chainId.toString(),
+    to: simulationParams.to,
+    input: simulationParams.input,
+    value: simulationParams.value,
+    from: simulationParams.from || AddressZero,
+    root: simulationParams.fork?.root,
+  };
+
+  if (simulationParams.timestampOverride !== undefined) {
+    body.block_header = {
+      timestamp: BigNumber.from(simulationParams.timestampOverride).toHexString(),
+    };
+  }
+
+  return body;
+};
+
+// Type guard function to check if the API response conforms to the required TenderlyAPIResponse interface
+function isTenderlyAPIResponse(response: any): response is TenderlyAPIResponse {
+  if (
+    response &&
+    response.simulation &&
+    typeof response.simulation.id === "string" &&
+    typeof response.simulation.status === "boolean"
+  ) {
+    return true;
+  }
+  return false;
+}
+
+const getSimulationResponse = async (
+  simulationParams: TenderlySimulationParams,
+  tenderlyEnv: TenderlyEnvironment,
+  retryOptions: RetryOptions
+): Promise<TenderlyAPIResponse> => {
+  // Construct Tenderly simulation API request.
+  const url = createRequestUrl(tenderlyEnv, simulationParams.fork);
+  const body = createRequestBody(simulationParams);
+  const headers = { "X-Access-Key": tenderlyEnv.apiKey };
+
+  // Send Tenderly simulation API request with retries.
+  const response = await retry(async () => {
+    const fetchResponse = await fetch(url, {
+      method: "POST",
+      headers: headers,
+      body: JSON.stringify(body),
+    });
+    if (!fetchResponse.ok) {
+      throw new Error(`Simulation API returned HTTP ${fetchResponse.status}: ${fetchResponse.statusText}`);
+    }
+    return fetchResponse;
+  }, retryOptions);
+
+  // If the HTTP response was OK, we expect the response body should be a JSON object containing expected Tenderly
+  // simulation response properties.
+  const apiResponse = await response.json();
+  if (!isTenderlyAPIResponse(apiResponse)) {
+    throw new Error(`Failed to parse Tenderly simulation API response: ${JSON.stringify(apiResponse)}`);
+  }
+  return apiResponse;
+};
+
+const isProjectPublic = async (tenderlyEnv: TenderlyEnvironment): Promise<boolean> => {
+  const url = `https://api.tenderly.co/api/v1/public/account/${tenderlyEnv.user}/project/${tenderlyEnv.project}`;
+
+  // Return true only if the project API responds OK and the project is public. On any error, return false.
+  try {
```

Comment:
Note: using a try catch here may cause false negatives.

A temporary API failure could cause this to return false, when it should be true. Is that resiliency issue okay for the expected application of this function? Or should we refactor this to detect the specific case of a private project and bubble up an error otherwise (or wrap in a retry)?

Reply:
To reduce false negatives, I now wrapped fetch in retry with retryOptions that can be passed by consuming application.
I also now added authentication headers, so that even private project responds correctly that it is private.
Still, if there are persistent API failures, the calling application should decide if it is ok with private simulation link.